### ISSUE TITLE
[FIRRTL] SFC Memory Lowering: Respect Write Collision Behavior

### ISSUE THREAD

**Issue by seldridge** [llvm/circt#1757]
The default lowering for memories in the Scala FIRRTL Compiler (SFC) introduces a write ordering if the memories have the ports have the same clock.  This creates problems when trying to formally verify SFC-generate Verilog with CIRCT-generate Verilog.

Add an option that merges ports with the same clock into the same always block where the writes are ordered based on the port order.  This should probably be done as an extension of the existing generator schema that sets an optional ordering of write ports.  During FIRRTL memory lowering, this flag would then set the ordering based on the declared order of FIRRTL memory write ports.

As an example, consider the following FIRRTL circuit:

```scala
circuit Foo:
  module Foo:
    input clock: Clock
    input r0: {addr: UInt<4>, en: UInt<1>, flip data: UInt<8>}
    input w0: {addr: UInt<4>, en: UInt<1>, data: UInt<8>, mask: UInt<1>}
    input w1: {addr: UInt<4>, en: UInt<1>, data: UInt<8>, mask: UInt<1>}

    wire _r0: {addr: UInt<4>, en: UInt<1>, clk: Clock, flip data: UInt<8>}
    _r0 <- r0
    _r0.clk <= clock

    wire _w0: {addr: UInt<4>, en: UInt<1>, clk: Clock, data: UInt<8>, mask: UInt<1>}
    _w0 <- w0
    _w0.clk <= clock

    wire _w1: {addr: UInt<4>, en: UInt<1>, clk: Clock, data: UInt<8>, mask: UInt<1>}
    _w1 <- w1
    _w1.clk <= clock

    mem memory:
      data-type => UInt<8>
      depth => 16
      reader => r0
      writer => w0
      writer => w1
      read-latency => 1
      write-latency => 1
      read-under-write => undefined

    memory.r0 <= _r0
    memory.w0 <= _w0
    memory.w1 <= _w1
```

The SFC will produce the following verilog (`firrtl -i Foo.fir && cat Foo.v | verilator -E -P /dev/stdin`):

```verilog
module Foo(
  input        clock,
  input  [3:0] r0_addr,
  input        r0_en,
  output [7:0] r0_data,
  input  [3:0] w0_addr,
  input        w0_en,
  input  [7:0] w0_data,
  input        w0_mask,
  input  [3:0] w1_addr,
  input        w1_en,
  input  [7:0] w1_data,
  input        w1_mask
);
  reg [7:0] memory [0:15];
  wire  memory_r0_en;
  wire [3:0] memory_r0_addr;
  wire [7:0] memory_r0_data;
  wire [7:0] memory_w0_data;
  wire [3:0] memory_w0_addr;
  wire  memory_w0_mask;
  wire  memory_w0_en;
  wire [7:0] memory_w1_data;
  wire [3:0] memory_w1_addr;
  wire  memory_w1_mask;
  wire  memory_w1_en;
  reg  memory_r0_en_pipe_0;
  reg [3:0] memory_r0_addr_pipe_0;
  assign memory_r0_en = memory_r0_en_pipe_0;
  assign memory_r0_addr = memory_r0_addr_pipe_0;
  assign memory_r0_data = memory[memory_r0_addr];
  assign memory_w0_data = w0_data;
  assign memory_w0_addr = w0_addr;
  assign memory_w0_mask = w0_mask;
  assign memory_w0_en = w0_en;
  assign memory_w1_data = w1_data;
  assign memory_w1_addr = w1_addr;
  assign memory_w1_mask = w1_mask;
  assign memory_w1_en = w1_en;
  assign r0_data = memory_r0_data;
  always @(posedge clock) begin
    if (memory_w0_en & memory_w0_mask) begin
      memory[memory_w0_addr] <= memory_w0_data;
    end
    if (memory_w1_en & memory_w1_mask) begin
      memory[memory_w1_addr] <= memory_w1_data;
    end
    memory_r0_en_pipe_0 <= r0_en;
    if (r0_en) begin
      memory_r0_addr_pipe_0 <= r0_addr;
    end
  end
initial begin
end  
endmodule
```

CIRCT produces the following which is not formally equivalent (`firtool Foo.fir -verilog`):

```verilog
module FIRRTLMem_1_2_0_8_16_1_1_0(	// Foo.fir:25:5
  input        ro_clock_0, ro_en_0,
  input  [3:0] ro_addr_0,
  input        wo_clock_0, wo_en_0,
  input  [3:0] wo_addr_0,
  input        wo_mask_0,
  input  [7:0] wo_data_0,
  input        wo_clock_1, wo_en_1,
  input  [3:0] wo_addr_1,
  input        wo_mask_1,
  input  [7:0] wo_data_1,
  output [7:0] ro_data_0);

  reg [7:0] Memory[0:15];
  reg       _T;
  reg [3:0] _T_0;

  always @(posedge ro_clock_0) begin
    _T <= ro_en_0;
    _T_0 <= ro_addr_0;
  end // always @(posedge)
  always @(posedge wo_clock_0) begin
    if (wo_en_0 & wo_mask_0)
      Memory[wo_addr_0] <= wo_data_0;
  end // always @(posedge)
  always @(posedge wo_clock_1) begin
    if (wo_en_1 & wo_mask_1)
      Memory[wo_addr_1] <= wo_data_1;
  end // always @(posedge)
  assign ro_data_0 = _T ? Memory[_T_0] : 8'bx;	// Foo.fir:25:5
endmodule

module Foo(	// Foo.fir:2:10
  input        clock,
  input  [3:0] r0_addr,
  input        r0_en,
  input  [3:0] w0_addr,
  input        w0_en,
  input  [7:0] w0_data,
  input        w0_mask,
  input  [3:0] w1_addr,
  input        w1_en,
  input  [7:0] w1_data,
  input        w1_mask,
  output [7:0] r0_data);

  FIRRTLMem_1_2_0_8_16_1_1_0 memory (	// Foo.fir:25:5
    .ro_clock_0 (clock),
    .ro_en_0    (r0_en),
    .ro_addr_0  (r0_addr),
    .wo_clock_0 (clock),
    .wo_en_0    (w0_en),
    .wo_addr_0  (w0_addr),
    .wo_mask_0  (w0_mask),
    .wo_data_0  (w0_data),
    .wo_clock_1 (clock),
    .wo_en_1    (w1_en),
    .wo_addr_1  (w1_addr),
    .wo_mask_1  (w1_mask),
    .wo_data_1  (w1_data),
    .ro_data_0  (r0_data)
  );
endmodule
```

The specific issue is that the SFC version is enforcing a write ordering where simultaneous writes to the `w0` and `w1` port will result in the `w1` port winning. In the CIRCT version, this is undefined behavior. Abstractly, and without any other information, the CIRCT generation is "better" as it leaves it up to the Synthesis tool to choose an ordering.  (E.g., FPGA tools prefer separate always blocks so that they don't have to code in an explicit ordering to align with what the SFC code is behaviorally describing.)

Note that If the write ports are reordered in the FIRRTL definition so that `w1` comes before `w0`, then you will get `w0` "winning" on collision

**Comment by seldridge** [comment]
With #1805, the above will be lowered to:

```verilog
  always @(posedge ro_clock_0) begin
    _T <= ro_en_0;
    _T_0 <= ro_addr_0;
  end // always @(posedge)
  always @(posedge wo_clock_0) begin
    if (wo_en_0 & wo_mask_0)
      Memory[wo_addr_0] <= wo_data_0;
    if (wo_en_1 & wo_mask_1)
      Memory[wo_addr_1] <= wo_data_1;
  end // always @(posedge)
```

If instead the second memory port was driven by a `clock2`, then you get split always blocks:

```verilog
  always @(posedge wo_clock_0) begin
    if (wo_en_0 & wo_mask_0)
      Memory[wo_addr_0] <= wo_data_0;
  end // always @(posedge)
  always @(posedge wo_clock_1) begin
    if (wo_en_1 & wo_mask_1)
      Memory[wo_addr_1] <= wo_data_1;
  end // always @(posedge)
```
